test/boost/reader_concurrency_semaphore_test.cc

```diff
@@ -1645,3 +1645,253 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_permit_waiting_for_me
     BOOST_REQUIRE_EQUAL(permit2.get_state(), reader_permit::state::evicted);
     BOOST_REQUIRE_THROW(res_fut.get(), std::bad_alloc);
 }
+
+// Check that inactive reads are not needlessly evicted when admission is not
+// blocked on resources.
+// This test covers all the cases where eviction should **not** happen.
+SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_no_unnecessary_evicting) {
+    const auto initial_resources = reader_concurrency_semaphore::resources{2, 4 * 1024};
+    const auto serialize_multiplier = std::numeric_limits<uint32_t>::max();
+    const auto kill_multiplier = std::numeric_limits<uint32_t>::max();
+    reader_concurrency_semaphore semaphore(initial_resources.count, initial_resources.memory, get_name(), 100,
+            utils::updateable_value<uint32_t>(serialize_multiplier), utils::updateable_value<uint32_t>(kill_multiplier));
+    auto stop_sem = deferred_stop(semaphore);
+
+    simple_schema ss;
+    auto s = ss.schema();
+
+    auto permit1 = semaphore.obtain_permit(nullptr, get_name(), 1024, db::no_timeout).get();
```

Comment:
in #13255, we [changed](https://github.com/scylladb/scylladb/pull/13255/commits/156e5d346d427abc4c79075cb2e71d68ed35d703#diff-46ce50f665ed19eaed28937148d633880d420c375fa7fbed2cb9261047a57ca9R389) the function signature to pass `trace_state` down to semaphore, so might want to update this caller accordingly.